### PR TITLE
Load tokens from HADOOP_TOKEN_FILE_LOCATION file and add to proxy user UGI for permission

### PR DESCRIPTION
**Pull request by tsangz2013**
**What need to be fixed:**
Yarn Client killApplication() call never succeeds in in containerization, as the proxyUser UGI never have proper token loaded.
And this RP in combination to #3164, the yarn application kill function works correctly in containerization.

**What is being done:**
When creating the proxyUser, use HadoopSecurityManager to get tokens from the previously stored tokenFile from file system, and grand them into the proxyUser.

**Tests done:**

- Tested with multiple different flows across different clusters, 
- kill yarn application works correctly, 
- verified in yarnUI that the call is being made from the pod as the proxyUser

### CODE REVIEW

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/commons/HadoopSecurityManager.java [hunk 1]
```diff
@@ -101,6 +102,9 @@ public abstract FileSystem getFSAsUser(String realIdentity, String proxyUser)
   public abstract void cancelTokens(File tokenFile, String userToProxy,
       Logger logger) throws HadoopSecurityManagerException;
 
+  public abstract Credentials getTokens(File tokenFile, Logger logger)
+      throws HadoopSecurityManagerException;
+
```
Comment (tsangz2013) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/commons/HadoopSecurityManager.java:107]: HSM expose a new method to get tokens from file, and implemented in AbstractHSM, which will be inherent by KsudoHSM.

**Review by tsangz2013 — COMMENTED** [review]
For better understanding

File: az-hadoop-jobtype-plugin/src/main/java/azkaban/jobtype/HadoopSecureWrapperUtils.java [hunk 2]
```diff
@@ -59,32 +66,61 @@ private static UserGroupInformation createSecurityEnabledProxyUser(String userTo
...
+          proxyUser));
+      for (Token<?> token : creds.getAllTokens()) {
+        proxyUser.addToken(token);
+        log.info(String.format("Token = %s, %s, %s ", token.getKind(), token.getService(),
+            Arrays.toString(token.getIdentifier())));
+      }
+      proxyUser.addCredentials(creds);
+    }
+
+    log.info("token copy finished for " + loginUser.getUserName());
```
Comment (svnarumugam) [inline az-hadoop-jobtype-plugin/src/main/java/azkaban/jobtype/HadoopSecureWrapperUtils.java:96]: The login user is `azdata`. proxy user is any proxy user P. wouldn't it be sufficient if we do this alone? I am trying to figure out why we are doing the copy from line 76 to line 81. am i missing something?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/AbstractHadoopSecurityManager.java [hunk 3]
```diff
@@ -401,12 +401,22 @@ protected void doPrefetch(final File tokenFile, final Props props, final Logger
...
 
     try {
+      // cred is being populated
       fetchAllHadoopTokens(userToProxyFQN, userToProxy, props, logger, cred);
       getProxiedUser(userToProxyFQN).doAs((PrivilegedExceptionAction<Void>) () -> {
         registerAllCustomCredentials(userToProxy, props, cred, logger);
         return null;
       });
+
+      logger.info("fetched cred = " + cred);
```
Comment (svnarumugam) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/AbstractHadoopSecurityManager.java:411]: [Logging] any strong reasons we shouldn't keep this as a `debug` level? I have the same question for other logging too.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/AbstractHadoopSecurityManager.java [hunk 4]
```diff
@@ -618,6 +628,16 @@ public void cancelTokens(final File tokenFile, final String userToProxy, final L
     }
   }
 
+  @Override
+  public Credentials getTokens(File tokenFile, Logger logger)
+      throws HadoopSecurityManagerException {
+    try {
+      return Credentials.readTokenStorageFile(new Path(tokenFile.toURI()), this.conf);
```
Comment (svnarumugam) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/AbstractHadoopSecurityManager.java:635]: we seem to be not using the logger instance here. did we intent to log something?

**Review by svnarumugam — CHANGES_REQUESTED** [review]
Functionally the changes look good to me. Thank you so much @tsangz2013 for working on this. I have left some comments.

File: az-hadoop-jobtype-plugin/src/main/java/azkaban/jobtype/HadoopSecureWrapperUtils.java (hunk 2 above)
Comment (tsangz2013) [reply az-hadoop-jobtype-plugin/src/main/java/azkaban/jobtype/HadoopSecureWrapperUtils.java:96]: I don't think using any user out of nowhere is good, plus this is the intend of `proxy`User, right?
Duplicate code but these are 2 sources of tokens, first is copy from login user, second is from token-file.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/AbstractHadoopSecurityManager.java (hunk 4 above)
Comment (tsangz2013) [reply azkaban-hadoop-security-plugin/src/main/java/azkaban/security/AbstractHadoopSecurityManager.java:635]: Good call, just follow the pattern for other HSM methods

**Review by svnarumugam — APPROVED** [review]
The changes look good to me. I think we can revisit the token copy part afterwards. The logging should be fine.

**Review by svnarumugam — APPROVED** [review]
Looks good to me.

**Pull request merged**